bee-network/src/service/service.rs

```diff
@@ -0,0 +1,486 @@
+    debug!("Command handler running.");
+
+    let mut commands = ShutdownStream::new(shutdown, UnboundedReceiverStream::new(commands));
+
+    while let Some(command) = commands.next().await {
+        if let Err(e) = process_command(command, &senders, &peerlist).await {
+            error!("Error processing command. Cause: {}", e);
+            continue;
+        }
+    }
+
+    debug!("Command processor stopped.");
+}
+
+async fn event_processor(shutdown: Shutdown, events: InternalEventReceiver, senders: Senders, peerlist: PeerList) {
+    debug!("Event handler running.");
+
+    let mut int_events = ShutdownStream::new(shutdown, UnboundedReceiverStream::new(events));
+
+    while let Some(int_event) = int_events.next().await {
+        if let Err(e) = process_internal_event(int_event, &senders, &peerlist).await {
+            error!("Error processing internal event. Cause: {}", e);
+            continue;
+        }
+    }
+
+    debug!("Event processor stopped.");
+}
+
+// TODO: implement exponential back-off to not spam the peer with reconnect attempts.
+async fn peer_checker(shutdown: Shutdown, senders: Senders, peerlist: PeerList) {
+    debug!("Peer checker running.");
+
+    let Senders { internal_commands, .. } = senders;
+
+    // NOTE:
+    // We add a random amount of milliseconds to when the reconnector starts, so that even if 2 nodes
+    // go online at the same time the probablilty of them simultaneously dialing each other is reduced
+    // significantly.
+
+    let delay = Duration::from_millis(rand::thread_rng().gen_range(0u64..1000));
+    let start = Instant::now() + delay;
+    let period = Duration::from_secs(reconnect_interval_secs()); // `unwrap` is safe!
+
+    let mut interval = ShutdownStream::new(shutdown, IntervalStream::new(time::interval_at(start, period)));
+
+    // Check, if there are any disconnected known peers, and schedule a reconnect attempt for each
+    // of those.
+    while interval.next().await.is_some() {
+        let peerlist = peerlist.0.read().await;
+
+        for (peer_id, alias) in peerlist.iter_if(|info, state| info.relation.is_known() && state.is_disconnected()) {
+            info!("Trying to reconnect to: {} ({}).", alias, alias!(peer_id));
+
+            // Ignore if the command fails. We can always try another time.
+            let _ = internal_commands.send(Command::DialPeer { peer_id });
+        }
+    }
+
+    debug!("Peer checker stopped.");
+}
+
+async fn process_command(command: Command, senders: &Senders, peerlist: &PeerList) -> Result<(), PeerError> {
+    trace!("Received {:?}.", command);
+
+    match command {
+        Command::AddPeer {
+            peer_id,
+            multiaddr,
+            alias,
+            relation,
+        } => {
+            let alias = alias.unwrap_or_else(|| alias!(peer_id).to_string());
+
+            // Note: the control flow seems to violate DRY principle, but we only need to clone `id` in one branch.
```

Comment:
I see no cloning going on here? Did `PeerId` become `Copy`? If so, this comment can be removed and this flow control simplified.

Reply:
yes, it became `Copy`.